segments_tree.go

```diff
@@ -0,0 +1,155 @@
+package quamina
+
+import (
+	"fmt"
+	"strings"
+)
+
+const SegmentSeparator = "\n"
+
+type segmentsTree struct {
+	root bool
+
+	// "nodes" stores a map from a segment to it's childrens
+	// in an hierarchial data like JSON, a node can be Object or Array.
+	// for example, in this path "context\nuser\nid", both "context" and "user" will be nodes.
+	nodes map[string]*segmentsTree
+
+	// "fields" is storing leaf fields here mapped to their full paths
+	// for example:
+	//  leaf "id" will be mapped to it's full path "context\nuser\nid"
+	//  leaf "user" will be mapped to it's full path "context\nuser"
+	fields map[string][]byte
+}
+
+// newSegmentsIndex creates a segmentsTree node which is the root
+// passing paths will auto-add them to the tree, useful for testing.
+func newSegmentsIndex(paths ...string) *segmentsTree {
+	st := newSegmentsIndexNode(true)
+
+	for _, path := range paths {
+		st.add(path)
+	}
+
+	return st
+}
+
+// newSegmentsIndexNode initializes a segmentsTree node
+func newSegmentsIndexNode(root bool) *segmentsTree {
+	return &segmentsTree{
+		root:   root,
+		nodes:  make(map[string]*segmentsTree),
+		fields: make(map[string][]byte),
+	}
+}
+
+func (p *segmentsTree) add(path string) {
```

Comment:
Thanks, much clearer now.  I still haven't 100% figured this out, but I think I'm starting to get it. The `[len-2]` was breaking my brain.

Reply:
Yeah, the `len-2` and skipping the first element was too complex.. this method can be re-written BTW, let me know what you think about it:

```go
	// Add the nodes
	for _, segment := range segments[:len(segments)-1] {
		node = node.getOrCreate(segment)
	}
	// Add the leaf
	node.addSegment(segments[len(segments)-1], []byte(path))
```